### PR TITLE
[python-api] - several additions

### PR DESCRIPTION
**Pull request by phil65**
- getCurrentContainerId() allows to properly use the core viewtype management for WindowXML instances.
- setInfo("dbid", xxx) allows to properly emulate database items. This is especially important for stuff like the dataprovider script from @BigNoid. Several context item add-ons depend on listitem.dbid for example.
- getArt() should also be self-explaining.  

@tamland

### CODE REVIEW

**Comment by ronie** [comment]
> allows to properly use the core viewtype management for WindowXML instances.

sounds cool... though i have no idea what it means. ;-)
care to elaborate?

**Comment by phil65** [comment]
Sure. Most of the time people add content to WindowXML lists via Control.addItems(), but there is also another way to do this, namely WindowXML.addItem(). When using that, you basically get MediaWindow behaviour for that list (example: you can reference to the listitems via listitem.xxx from outside of the container).
The problem is that the script author does not know which viewtype is currently active. By adding getCurrentContainerId(), that will become possible.

**Comment by phil65** [comment]
I appended another commit. While dealing with the new dbid setters and getters i noticed that items with a VideoinfoTag return -1 as dbid in case they dont have a dbid assigned. I think for skinners it is more convenient to check with IsEmpty() instead of isGreater(-1) for dbids. Agreed, @ronie ?

**Comment by Montellese** [comment]
Doesn't the last commit break any skin comparing against `-1`?

Concerning the possibility for plugins to set the database id this might have much further implications than you might think. A lot of code uses the fact that an item has a database id to differentiate between library and plugin items. With this change that is not always true anymore.
Could you explain the exact use case how this would be used? A plugin must not set an arbitrary database id on an item that is not stored in the video or music database or some core logic ends up thinking that the item is an actual library item and overwrites the information in the database with the one from the plugin which might not be wanted.

**Comment by ronie** [comment]
> I think for skinners it is more convenient to check with IsEmpty() instead of isGreater(-1) for dbids. Agreed, @ronie ?

for sure, if no info is available it should return empty.

**Comment by phil65** [comment]
> Doesn't the last commit break any skin comparing against -1?

In theory yes, but breakage will be minimal. We have much bigger breakers in v17. :)
For me issues mainly came up when doing script work (I am enriching remote metadata with local metadata if available in my add-ons)

> Could you explain the exact use case how this would be used? A plugin must not set an arbitrary database id on an item that is not stored in the video or music database or some core logic ends up thinking that the item is an actual library item and overwrites the information in the database with the one from the plugin which might not be wanted.

A valid usecase for plugins (esp. for those which get integrated into skins via dynamic content) is to imitate library listitems. A good example probably is perhaps service.library.data.provider, which provides listitems for several widget types. Those scripts have to do hacks like https://github.com/BigNoid/service.library.data.provider/blob/master/default.py#L269 for example in order to allow skinners or other scripts to have the correct dbid at hand. (example: https://github.com/phil65/context.extendedinfo.dialog/blob/krypton/addon.py#L28-L29 )
This PR tries to get rid of those hacks.

**Comment by Montellese** [comment]
@phil65: I've never used `service.library.data.provider` or any of the other add-ons you mentioned so excuse my ignorance. Are the imitated library items based on real library items just with additional information? Or ar they fake items that don't exist in the database?

**Comment by phil65** [comment]
They are based on real library items. The aim for those plugins basically is to create listitems via python which are identical to the ones delivered via core, all infolabels should be identical.

File: xbmc/interfaces/legacy/ListItem.h [hunk 1]
```diff
@@ -333,6 +333,17 @@ namespace XBMCAddon
...
       /**
+       * getArt(key) -- Returns a listitem art value as a string, similar to an infolabel.\n
+       * \n
+       * key            : string - art name.\n
+       * \n
+       *
+       * example:
+       *   - poster = self.list.getSelectedItem().getArt('poster')
+       */
+      String getArt(const char* key);
```
Comment (tamland) [inline xbmc/interfaces/legacy/ListItem.h]: Could use some better documentation. 'It returns a value'  A value of what? And list the available types please, as setArt does.

**Comment by Montellese** [comment]
How do we protect core against add-ons using wrong/arbitrary database ids which could either overwrite data from existing library items or try to write library items that don't exist?

**Comment by phil65** [comment]
In my opinion a warning in docs should be fine that this should only be used for local items and in combination with the correct mediatype value, I can add that. Would that be sufficient for you?

File: xbmc/interfaces/legacy/ListItem.h (hunk 1 above)
Comment (phil65) [inline xbmc/interfaces/legacy/ListItem.h]: Ok.

**Comment by Montellese** [comment]
Ideally there would be a way to prevent plugins from using invalid database ids in the python interface implementation but that's probably not possible right now or would be too much work so a warning is certainly nice.
Maybe we should change the method to set the id and the type to force plugin authors to set both at the same time.

**Comment by phil65** [comment]
@Montellese : I updated the docs, is it fine enough now?

**Comment by Montellese** [comment]
Yes, better.

**Comment by phil65** [comment]
jenkins build this please.

**Comment by mapfau** [comment]
This PR leads to an issue commented in:

https://github.com/xbmc/xbmc/commit/65ce5ef4753e3a3d12577bd487a089adf9912479

**Comment by phil65** [comment]
Already fixed on estuary repository.

**Comment by phil65** [comment]
https://github.com/xbmc/skin.estuary/commit/d13f4c935c9dea4525c7bea9a8393b4e8ea1fc69

**Comment by tamland** [comment]
@phil65 you should push things like that to _this_ repo because otherwise it will be left broken in nightly and for everyone doing development. Also, that's a private repo mapfau can't access.

**Comment by phil65** [comment]
We´re in pre-alpha, I noticed it after the merge, and there are also a lot of other small fixes I dont push on a daily basis....

**Comment by tamland** [comment]
Sure, you don't have to push every small fix daily. But you can see how it's not ideal that issues are known and fixes exists, but can't be acquired yet right?

**Comment by phil65** [comment]
Then the thing to do is probably to discuss whether the estuary repo should get public, and not instantly attacking someone for their workflow (which was agreed on by the team).

**Comment by phil65** [comment]
https://github.com/xbmc/xbmc/pull/9665

**Comment by tamland** [comment]
It was never my intention to attack anyone. sorry for bringing it up

**Pull request merged**